### DISCUSSION TITLE
Reorganize the Fixpoint Engine?

### DISCUSSION THREAD

**Discussion by stephentetley** [flix/flix#6935] [General]
Should the fixpoint engine be reorganized to be more like the Flix compiler itself? 

At the moment the fixpoint engine doesn't really use namespaces and the file layout arguably makes harder to see how the engine works. I propose merging the Datalog and Ram types into larger units under in an `Ast` namespace like Flix compiler:

~~~

Ast.Datalog
Ast.PrecedenceGraph
Ast.Ram
Ast.Shared      // PredSym & Denotation

~~~

`PrecedenceGraph` seems to merit its own module and `PredSym` and `Denotation` are used by both `Datalog` and `Ram` types.

Likewise I think a `Phase` namespace would help show how the engine's processing works with the modules called by `Solver.solveWithStratification` moved into this namespace:

~~~

Phase.Compiler
Phase.IndexSelection
Phase.Simplifier
Phase.Stratifier
Phase.VarsToIndices

~~~

Any thoughts?

**Comment by magnus-madsen** [comment]
Hi Stephen,

I agree with your suggestions. Do you want to have a go at it? 

On a related note, I have a Msc student who will be working on the Datalog aspect in Flix. He starts in about a month. The project itself is yet to be determined.

**Comment by stephentetley** [comment]
Thanks Magnus - I'll make a PR.

**Reply by magnus-madsen** [reply]
Possibly a few smaller PRs would be useful.
